### PR TITLE
LibJS: Clean up how parser errors work

### PR DESCRIPTION
**Pull request by linusg**
Addresses https://github.com/SerenityOS/serenity/issues/2207, I guess.

- LibJS: Remove syntax errors from lexer 

  Giving the lexer the ability to generate errors adds unnecessary complexity - also it only calls its `syntax_error()` function in one place anyway ("unterminated string literal"). But since the lexer *also* emits tokens like `Eof` or `UnterminatedStringLiteral`, it should be up to the consumer of these tokens to decide what to do.

  Also remove the option to not print errors to stderr as that's not relevant anymore.

- LibJS: Let parser keep track of errors

  Rather than printing them to stderr directly the parser now keeps a `Vector<Error>`, which allows the "owner" of the parser to consume them individually after parsing.

  The `Error` struct has a message, line number, column number and a `to_string()` helper function to format this information into a meaningful error message.

  The `Function()` constructor will now include an error message when throwing a `SyntaxError`.

- js: Throw a regular SyntaxError for errors from the parser

  ```
  > foo]bar)baz
  Uncaught exception: [SyntaxError]: Unexpected token BracketClose. Expected Semicolon (line: 1, column: 5)
  >
  ```

### CODE REVIEW

File: Libraries/LibJS/Lexer.h [hunk 1]
```diff
@@ -37,14 +38,9 @@ namespace JS {
 class Lexer {
 public:
     explicit Lexer(StringView source);
-    Lexer(StringView source, bool log_errors)
-        : Lexer(source)
-    {
-        m_log_errors = log_errors;
-    }
+    // explicit Lexer(StringView source, AK::Function<void(String message, size_t line, size_t column)> on_error);
```
Comment (Dexesttp) [inline Libraries/LibJS/Lexer.h:41]: With the new error-free lexer, this won't be needed anymore and should probably be removed. I guess that it was leftover from the lexer refactoring ?

Comment (linusg) [reply Libraries/LibJS/Lexer.h:41]: Yep, that was me playing around with a lexer-just-reports-error-someone-else-takes-care-of-it approach. Thanks!

**Comment by sunverwerth** [comment]
image

**Comment by alimpfard** [comment]
@sunverwerth 
image

**Pull request merged**